src/shady-css/parser.js

```diff
@@ -0,0 +1,229 @@
+import { matcher, tokenType } from './common';
+import { TokenCursor } from './token-cursor';
+import { Lexer } from './lexer';
+import { NodeFactory } from './node-factory';
+
+class Parser {
+  constructor(nodeFactory = new NodeFactory()) {
+    this.nodeFactory = nodeFactory;
+  }
+
+  parse(cssText) {
+    return this.parseStylesheet(new TokenCursor(new Lexer(cssText)));
+  }
+
+  parseStylesheet(tokenCursor) {
+    let declarations = [];
+    let declaration;
+
+    while (tokenCursor.next &&
+           (declaration = this.parseDeclaration(tokenCursor))) {
+      declarations.push(declaration);
+    }
+
+    return this.nodeFactory.stylesheet(declarations);
+  }
+
+  parseDeclaration(tokenCursor) {
+    // Consume all leading (presumably insignificant) whitespace:
+    this.traverseWhitespace(tokenCursor);
+
+    if (tokenCursor.next) {
+      // Test for a comment opening. Take a comment as necessary:
+      if (matcher.commentOpen.test(tokenCursor.next.text)) {
```

Comment:
as we discussed about removing regex.test in parsing: this should be something like `if (token.type === TokenType.commentOpen)`